test/unit/Input.test.js

```diff
@@ -0,0 +1,105 @@
+import Input from '../../src/components/Input.react.js';
+import React from 'react';
+import {mount, render} from 'enzyme';
+
+test('Input renders', () => {
+    const input = render(<Input />);
+
+    expect(input.html()).toBeDefined();
+});
+
+describe('Props can be set properly', () => {
+    const defaultProps = {
+        id: 'input-1',
+        value: 'hello, dash',
+        style: {backgroundColor: 'hotpink'},
+        className: 'input-class',
+        type: 'text',
+        autoComplete: 'on',
+        autoFocus: 'on',
+        disabled: true,
+        inputMode: 'verbatim',
+        list: 'hello',
+        max: '2',
+        maxLength: '2',
+        min: '1',
+        minLength: '1',
+        n_blur: 0,
+        n_blur_timestamp: -1,
+        n_submit: 0,
+        n_submit_timestamp: -1,
+    };
+    const input = mount(<Input {...defaultProps} />);
+
+    test('props are being set', () => {
+        expect(input.props()).toBeDefined();
+        expect(input.props()).toEqual(defaultProps);
+    });
+
+    test('props.id is set as the <input> id', () => {
+        expect(input.props().id).toBeDefined();
+        expect(input.props().id).toEqual(defaultProps.id);
+
+        // test if id is in the actual HTML string
+        expect(
+            input
+                .render()
+                .children()
+                .attr('id')
+        ).toEqual(defaultProps.id);
+    });
+    test('props.value is set as the <input> value', () => {
+        expect(input.props().value).toBeDefined();
+        expect(input.props().value).toEqual(defaultProps.value);
+
+        // test if id is in the actual HTML string
+        expect(
+            input
+                .render()
+                .children()
+                .attr('value')
+        ).toEqual(defaultProps.value);
+    });
+    test('props.className is set as the <input> CSS class', () => {
+        expect(input.props().className).toBeDefined();
+        expect(input.props().className).toEqual(defaultProps.className);
+
+        // test if className is actually set on HTML output
+        expect(
+            input
+                .render()
+                .children()
+                .attr('class')
+        ).toEqual(defaultProps.className);
+    });
+});
+
+test('Input does not change state if it rerenders', () => {
+    const input = mount(<Input value="initial value" />);
+    expect(input.find('input').getNode().value).toEqual('initial value');
+
+    input.find('input').simulate('change', {target: {value: 'new value'}});
+
+    expect(input.find('input').getNode().value).toEqual('new value');
+
+    // dash-renderer could rerender a component with it's original
+    // props, if dash-renderer is not informed of prop updates
+    input.setProps({value: 'initial value'});
+
+    expect(input.find('input').getNode().value).toEqual('new value');
```

Comment:
Could you explain what is being tested here?

Reply:
Yes! In some cases, `dash-renderer` will need to re-render the entire layout. It will, at that point, push the props in the components it wants to render. If, Dash isn't informed of any prop updates (in the case were `setProps()` is not being used, but the component's internal state is) then it will push in the old prop, or the initial prop.